### ISSUE TITLE
Compile deadlock with ActiveJob

### ISSUE THREAD

**Issue by paulhenri-l** [ViewComponent/view_component#1488]
I think the fix to this issue #1447 has created a regression because I am now experiencing deadlocks when using view_component and turbo streams.

After a lot of digging I found out that ViewComponent may reset the CompileCache while a request running in another thread is rendering some ViewComponents.

The line that refreshes the cache is this one and It seems to be triggered right after action cable streams a message (So probably after the BoradcastJob is performed).
https://github.com/github/view_component/blob/6397e87fcd9e606212493f31127943e13e3e336b/lib/view_component/engine.rb#L136

The fact that the cache is reset makes the rendering on the other thread to recompile the view component event though it is already compiled. Because of that it tries to get a write lock while still holding a read lock thus resulting in a deadlock.
(If it already holds a read lock that is because my components are nested)

To sum it up

```
In thread #1
Send turbo stream message
Finish BroadcastJob
Reset CompileCache

In thread #2
Render view component A <-- (Compile and is now cached)
  Render view component A <-- (Knows it is compiled so renders with read lock)
    -- Thread #1 has reset cache --
    Render view component A <-- (Thinks it needs to compile and try to get write lock)
      deadlock because we already have a read lock.
```

### Steps to reproduce
Unfortunately that issue is extremely hard to reproduce because of its nature

### Expected behavior
No deadlock

### Actual behavior
Deadlock :(

**Comment by paulhenri-l** [comment]
It looks to me that the most straightforward way to fix this would be to use a different cache per thread

**Comment by allmarkedup** [comment]
I just wanted to dive in here quickly to say that I've been trying to debug [an issue with Lookbook](https://github.com/allmarkedup/lookbook/issues/135) whereby the server is intermittently hanging. From reading the description here and in #1447 it seems like it might well be caused by this deadlock issue (or a subtle variation of it).

In Lookbook I believe the problem is especially acute because of the way it uses websocket notifications to trigger a fetch of updated HTML from the server after a change has been made, with the HTML in part generated using a number of 'fake' requests (`controller.process` calls) to the ViewComponent preview controller, which in turn renders the components themselves. All a bit convoluted(!) but the end result is that it is very easy to get into a situation where a few quick changes to a preview can result in multiple parallel requests being made to render a component, and this seems to cause the server to hang fairly often.

The problem goes away when setting the `RAILS_MAX_THREADS` environment variable to `1` so it definitely appears to be thread related.

I've done some quick informal testing using @paulhenri-l's PR (#1489) and from what I can tell it does _seem_ to fix the issue, although I haven't tested very thoroughly so can't say for sure yet.

This level of Ruby is way above my pay grade so I'm afraid I can't contribute much to the discussion around the fix itself but if I can help test or with anything else please let me know as I'd love to see this issue get sorted.

**Comment by Spone** [comment]
I was experiencing the same issue when working locally with the latest versions of ViewComponent and Lookbook. I was wondering if the issue was related to my project's setup. I'm kind of relieved to learn that this is something others are experiencing as well, I hope we can work together to figure out what the culprit is.

**Comment by allmarkedup** [comment]
@spone yes it's been driving me a little bit crazy trying to get to the root cause of it so I was glad to come across this issue too! Hopefully I haven't been too hasty in assuming this is the cause of the problems in Lookbook but it certainly seems likely to be related.

**Comment by allmarkedup** [comment]
FWIW the hanging server issues I'm seeing also seem to be resolved by downgrading ViewComponent to <= 2.63.0. As 2.64.0 was the first version that included the fix to #1447 I suspect that it is indeed related to that (as @paulhenri-l suggested above).

**Comment by stillhart** [comment]
I think I run into the same issue but with some parallel json requests; no ActiveJob, no Turbo Streams. In my case I had a recursive ViewComponent which might render a ViewComponent inside a ViewComponent, this 100's of times. I was refactoring a recursive ApplicationHelper method into ViewComponents.

**Comment by paulhenri-l** [comment]
Well in my case I have been using my fork since then and have had no issue whatsoever so maybe we should work on trying to get this one merged :)

**Comment by nicbet** [comment]
FYI, we've been encountering this bug **outside** of Logbook, in our two of production apps:

1. the first application is a "reactive application" in that we asynchronously re-render parts of the UI in response to events triggered by background jobs processing and sending the UI updates through via Turbo stream updates; the UI pieces that get updated are ViewComponents. A page view while a background job render is in progress appears to deadlock the server in line with the bug reported in this issue.

2. the second application uses a "code example" ViewComponent containing two panels, one with the source code of a given ERB template, and another with what the rendered ERB template will look like. To do so it reads the ERB template contents with `File.read()`. Hitting the browser refresh button twice will cause a deadlock in line with the bug reported in this issue.

In both instances downgrading `view_component` to `2.63.0` does indeed resolve the server deadlocks. 👍

**Comment by camertron** [comment]
Hey everyone! Thanks for all the discussion around the current threadsafety issues. I just submitted a [pull request](https://github.com/ViewComponent/view_component/pull/1550) that removes the locking mechanisms entirely (see the PR description for my reasoning). I'd be grateful if folks could try it in their apps and report back.

**Comment by kirillplatonov** [comment]
@camertron tested it on my app. Works great. No more deadlocks! Thank you! 👏

**Comment by nicbet** [comment]
@camertron From (limited) testing earlier today - the code on that branch works great, no deadlocks so far! 👍 

Would be interested to understand the downsides of removing the locking mechanism for concurrency/thread safety/production apps.

**Comment by stevegeek** [comment]
@camertron Seems to have stopped the regular lockups for me too! Thanks!

**Comment by BlakeWilliams** [comment]
> Would be interested to understand the downsides of removing the locking mechanism for concurrency/thread safety/production apps.

We chatted about this a good bit internally. The core problem we were originally working around was the Ruby interpreter emitting a warning when you re-define a method.

e.g.

```ruby
obj.define(:call) do
  # some code
end

# Method is already defined above, so Ruby emits a warning
obj.define(:call) do
  # some code
end
```

To get around this, we were un-defining the method right before we re-define it. This is where we ran into concurrency issues since one thread could un-define `call` right before another thread called the `call` method (resulting in the `undefined method call` errors).

Now we use `silence_errors` when [re-defining the method](https://github.com/ViewComponent/view_component/pull/1550/files#diff-fa28fa6e2d5f267384e7793b855281451a46b8437431867871a298fc52fe4940R69) so the error message is no longer a concern, and we can avoid locks since we can override the existing method in place.

I'm a bit surprised production apps ran into this at all, and I'm really curious what settings they're running with since we shouldn't have any components being compiled during run-time. I also wouldn't expect the cache to be invalidated in production, so I'm curious if there's still an issue here.

If folks ran into this in production, I'd love to know if [this line](https://github.com/ViewComponent/view_component/blob/57aeb20465a88e0b8da405758b5e9d658c55121e/lib/view_component/engine.rb#L134-L136) of code is being hit, which is the primary place we invalidate template caches. Although I suppose it's possible it could be some unexpected fork/threading behavior too.

**Comment by allmarkedup** [comment]
As the others have said, this is working great for me, no more issues. Thanks for your work on this @camertron, much appreciated!

**Comment by paulhenri-l** [comment]
@BlakeWilliams In my case the issue only happened in development and never in production.
@camertron I'll try your pull request when I'll be back at work and keep you updated 👍

**Comment by Spone** [comment]
[v2.74.1](https://github.com/ViewComponent/view_component/releases/tag/v2.74.1) has been released, please try it out!

**Comment by joelhawksley** [comment]
Closing as fixed, please don't hesitate to comment here if you run into this issue again ❤️
